### ISSUE TITLE
Concering Inconsistent Results Based On * Operator

### ISSUE THREAD

**Issue by mlpotter** [google/jax#13042]
### Description

I raised a [github issue](https://github.com/pyro-ppl/numpyro/issues/1488) in the NumPyro github pages that there was a drastic incorrect difference in  MCMC inference results depending on the use of * operator, and was redirected to the jax folk. As you may see below, the exact same mathematical operators (factorized versus unfactorized) create serious problems for gradients affecting Hamiltonian MCMC sampling. I am a bit suprised simple arithmetic difference like this causes such problems.

> @mlpotter It is better to raise those numerical issues to jax folks. You can just create a function with arguments are latent variables and try to see if there are differences in the evaluation. You use the following code to make an issue in jax.
> 
> ```python
> import jax.numpy as jnp
> 
> t1 = jnp.array([0., 1.])
> t2 = jnp.array([1., 2.])
> 
> def f1(k, r):
>     lam = jnp.power(-jnp.log(r), 1 / k)
>     p = jnp.exp(-(jnp.power(t2*lam, k) - jnp.power(t1*lam, k)))
>     return p.sum()
> 
> def f2(k, r):
>     lam = jnp.power(-jnp.log(r), 1 / k)
>     p = jnp.exp(-(jnp.power(t2, k) - jnp.power(t1, k)) * jnp.power(lam, k))
>     return p.sum()
> 
> print(jax.device_get(jax.value_and_grad(f1, (0, 1))(0.1, 0.9)))
> print(jax.device_get(jax.value_and_grad(f2, (0, 1))(0.1, 0.9)))
> ```
> 
> which returns
> 
> ```
> (array(1.89246643), (array(nan), array(nan)))
> (array(1.89246643), (array(-0.07768232), array(1.0791475)))
> ```
I believe it is coming from how a log(0) is being handled in the gradient, but f2 handles this fine and f1 fails.

### What jax/jaxlib version are you using?

jaxv 0.3.20, jaxlibv 0.3.21

### Which accelerator(s) are you using?
cpu

### Additional system info

Edition	Windows 10 Home
Version	21H2
Installed on	‎12/‎8/‎2021
OS build	19044.2006
Experience	Windows Feature Experience Pack 120.2212.4180.0

### NVIDIA GPU info
+-----------------------------------------------------------------------------+
| NVIDIA-SMI 522.25       Driver Version: 522.25       CUDA Version: 11.8     |
|-------------------------------+----------------------+----------------------+
| GPU  Name            TCC/WDDM | Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|                               |                      |               MIG M. |
|===============================+======================+======================|
|   0  NVIDIA GeForce ... WDDM  | 00000000:01:00.0  On |                  N/A |
| N/A   51C    P8     9W /  N/A |   1613MiB /  8192MiB |      2%      Default |
|                               |                      |                  N/A |
+-------------------------------+----------------------+----------------------+

+-----------------------------------------------------------------------------+
| Processes:                                                                  |
|  GPU   GI   CI        PID   Type   Process name                  GPU Memory |
|        ID   ID                                                   Usage      |
|=============================================================================|
|    0   N/A  N/A      1532    C+G   ...e\root\Office16\EXCEL.EXE    N/A      |
|    0   N/A  N/A      5940    C+G   ...wekyb3d8bbwe\Video.UI.exe    N/A      |
|    0   N/A  N/A      7152    C+G   C:\Windows\explorer.exe         N/A      |
|    0   N/A  N/A     10432    C+G   ...5n1h2txyewy\SearchApp.exe    N/A      |
|    0   N/A  N/A     10592    C+G   ...2txyewy\TextInputHost.exe    N/A      |
|    0   N/A  N/A     11036    C+G   ...artMenuExperienceHost.exe    N/A      |
|    0   N/A  N/A     11436    C+G   ...5n1h2txyewy\SearchApp.exe    N/A      |
|    0   N/A  N/A     13388    C+G   ...cw5n1h2txyewy\LockApp.exe    N/A      |
|    0   N/A  N/A     14824    C+G   ...oot\Office16\POWERPNT.EXE    N/A      |
|    0   N/A  N/A     19884    C+G   ...2gh52qy24etm\Nahimic3.exe    N/A      |
|    0   N/A  N/A     22196    C+G   ...lPanel\SystemSettings.exe    N/A      |
|    0   N/A  N/A     24464    C+G   ...y\ShellExperienceHost.exe    N/A      |
|    0   N/A  N/A     28308    C+G   ...oot\Office16\POWERPNT.EXE    N/A      |
|    0   N/A  N/A     29336    C+G   ...a\bin\win64\cefhelper.exe    N/A      |
|    0   N/A  N/A     36180    C+G   ...022a\bin\win64\MATLAB.exe    N/A      |
|    0   N/A  N/A     36512    C+G   ...ge\Application\msedge.exe    N/A      |
|    0   N/A  N/A     37892    C+G   ...perience\NVIDIA Share.exe    N/A      |
+-----------------------------------------------------------------------------+

**Comment by jakevdp** [comment]
Thanks for the report – I dug into this a bit, and I suspect this is the issue at play here:
```python
import jax
import jax.numpy as jnp

t = 0.0

def f1(r, k):
  return jnp.power(t * r, k)

def f2(r, k):
  return jnp.power(t, k) * jnp.power(r, k)

print(jax.grad(f1)(1., 0.1))
# nan
print(jax.grad(f2)(1., 0.1))
# 0.0
```
You can maybe understand why this is happening by thinking about applying the chain rule to the two expressions. For the first we get something like this:

- $f1(x) = (0 \cdot x)^k$
- $df1/dx = k \cdot 0 \cdot (0x)^{k - 1}$
- $df1/dx = 0.1 \cdot 0 / 0^{0.9}$
- $df1/dx = NaN$

For the second we get something like this:

- $f2(x) = 0^k \cdot x^k$
- $df2/dx = 0^k \cdot k \cdot x^{k - 1}$
- $df2/dx = 0 \cdot 0.1 / x^{0.9}$
- $ df2/dx = 0$

That is, even when computing the derivatives of this kind of expression purely symbolically, the order of operations actually matters in terms of whether the gradient is defined!

I'm not sure what the resolution is here, but I hope that shines some light on what might be going on.

**Comment by mlpotter** [comment]
Yes the chain rule derivatives make sense from the purely symbolic standpoint, it is just odd that somewhere/somehow RSTAN accounts for this during MCMC inference. 

It seems like if you have $(0*x)^k$ that that automatic differentiation package should prune this computation from the backpropagation graph because of course the derivative should be zero wrt x when this happens.

**Comment by jakevdp** [comment]
Thanks - agreed that autodiff should be able to handle this case correctly. Let me think about it a bit...

**Comment by jakevdp** [comment]
So, digging into this a little more, and unfortunately I think it will have to be marked as "working as intended". Given the mechanism behind JAX's autodiff system (and other autodiff systems in general), it's generally problematic if your computation includes units that produce infinities. In your case, you have a computational unit that produces an infinity as part of your expression (namely, `jnp.power(t1 * x, k)` when `0 < k < 1` and `t1 == 0`)

Identifying this problem within a large computation can be tricky, but JAX does provide the `jax_debug_nans` and `jax_debug_infs` flags that can be used to better isolate exactly where the problem comes up.

In this case, there are three possible approaches you might use:

1. re-express your computation in a way that avoids producing infinities. That's what your `f2` effectively does
2. define a [custom derivative rule](https://jax.readthedocs.io/en/latest/notebooks/Custom_derivative_rules_for_Python_code.html) for the particular combination of values that is leading to the problem. For example, you could have `def mul_pow(t, x, k): return jnp.power(t * x, k)` and define a custom jvp/vjp that correctly handles the case of $t = 0$ for $0 < k < 1$
3. Within the computation, use the [double-where](https://jax.readthedocs.io/en/latest/faq.html#gradients-contain-nan-where-using-where) trick to avoid the problematic value; for example you could replace `power(t * x, k)` with `where(t == 0, 0, power(t * x, k)`

Which one you use probably depends on your particular application. Hope that helps!
